File: Leetcode/Medium/221. Maximal Square.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        int res = 0;
        vector<vector<int>> dp(m+1, vector<int>(n+1,0));
        for(int i = 1; i < m+1; i++){
            for(int j = 1; j < n+1; j++){
                if(matrix[i-1][j-1] == '1'){
                    dp[i][j] = min(dp[i-1][j], min(dp[i-1][j-1], dp[i][j-1])) + 1;
                    res = max(res,dp[i][j] * dp[i][j]);
                }
            }
        }
        return res;
    }
};

File: Leetcode/Medium/221. Maximal Square/221. Maximal Square.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        """
        1 1 1
        1 2 1 

        """

        m, n = len(matrix), len(matrix[0])

        dp = [[0] * (n + 1) for _ in range(m + 1)]
        
        res = 0
        for r in range(1, m + 1):
            for c in range(1, n + 1):
                if matrix[r-1][c-1] == '1':
                    dp[r][c] = min(dp[r][c-1], dp[r-1][c-1], dp[r-1][c]) + 1
                res = max(dp[r][c] * dp[r][c], res)

        return res        